docs(uipath-data-fabric): filter operator support matrix + unsupported-operator decision flow (#1077)
UiPath/skills: skills/uipath-data-fabric/references/filter-platform-contract.md

# Filter Platform Contract

> Which operators are valid per field type, so you build a valid `records query` filter. Body usage and the unsupported-operator handling: [`records-query.md`](records-query.md) and SKILL.md Rule 17.

Reviewer: Sentence has grammatical issues, please check

## Filter body

```json
{
  "logicalOperator": "AND",          // AND/OR or 0/1 — case-insensitive
  "queryFilters": [
    { "fieldName": "Status", "operator": "=",  "value": "Active" },
    { "fieldName": "Status", "operator": "in", "valueList": ["A", "B"] }
  ],

>   "filterGroups": [ /* nested groups, recursive; AND/OR may mix per level */ ]

Reviewer: maybe mention that every nested group has the same structure as the parent?

>   "filterGroups": [ /* nested groups, recursive; AND/OR may mix per level */ ]

Reviewer: It isn't clear that this is a worked out example - I've seen this cause the agents confusion around what is the general contract, and what is an example - otherwise it incorrectly assumes certain fields' values are a must

Reviewer: Maybe we should declare the structure first, and then provide an example

## Operator support by field type

Build only within this matrix (✅ supported). The API *runs* some ❌ cells anyway (e.g. `<` on Text — lexicographic, so `"user2@…" < "user20@…"`) and 400s only on unknown operators (`==`, `Equals`, `like`). Never rely on that: when a request needs an unsupported operator/type combo, or has no value, ask the user — don't silently run it (SKILL.md Rule 17).

| Operator | Text / Multiline | Number / Autonum | Date/Time | Boolean | Choice Set | Relationship | File | Unique ID |
|---|---|---|---|---|---|---|---|---|
| `=` `!=` | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ❌ | ✅ |
| `contains` `not contains` | ✅ | ✅ | ❌ | ❌ | ✅ | ❌ | ❌ | ✅ |
| `startswith` `endswith` | ✅ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ | ✅ |
| `>` `<` `>=` `<=` | ❌ | ✅ | ✅ | ❌ | ❌ | ❌ | ❌ | ✅ |
| is empty / not empty | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| `in` `not in` | ✅ | ✅ | ✅ | ❌ | ❌ | ✅ | ❌ | ✅ |

Complex-field values: **Choice Set** — the integer `NumberId` (multi: `=` takes a sorted JSON-array string `"[1,3]"`, `contains` takes a bare id `"3"`). **Relationship** — the target record's UUID `Id`.

## Unsupported operator, or missing value

> If a request needs an out-of-matrix operator/type combo (or an operator outside the list above — `BETWEEN`, regex, `like`), or an operator other than is-empty/not-empty has no value, **don't silently run it**. Ask the user to either **(a)** run the query without that filter, or **(b)** supply a supported one — then apply only their choice, never a default. Compositions often help: `BETWEEN x AND y` → `>=` + `<=` in one `queryFilters` (`logicalOperator: 0`); regex → `contains` / `startswith` / `endswith`.

Reviewer: Necessary callout, this is a placeholder until we build and integrate `validate` tool right?

Author: yes few rules we can move to validate  like filter specific rules. Entity field naming

# Filter Platform Contract

> Which operators each field type accepts, so you can build a valid `records query` filter. For body usage and unsupported-operator handling, see [`records-query.md`](records-query.md) and SKILL.md Rule 17.

Reviewer: ```suggestion
Use this reference to find which operators are valid for each field type when building a `records query` filter. For body usage and unsupported-operator handling, see [`records-query.md`](records-query.md) and SKILL.md Rule 17.
```
